### PR TITLE
Modules service to aggregate module-level metadata and factories

### PR DESCRIPTION
**Pull request by dvoytenko**
Partial for #3742.

### CODE REVIEW

File: src/runtime.js [hunk 1]
```diff
@@ -173,7 +148,18 @@ function adoptShared(global, callback) {
...
   // Run specific setup for a single-doc or shadow-doc mode.
-  callback(global);
+  callback(global, modules);
+
+  /**
+   * @param {function(!Object)|{n:string, f:function(!Object)}} fnOrStruct
+   */
+  function installModule(fnOrStruct) {
+    if (typeof fnOrStruct == 'function') {
+      fnOrStruct(global.AMP);
```
Comment (cramforce) [inline src/runtime.js]: So, this is backwards compatible for modules as well? During roll out you need to expect extensions to use the old syntax.

File: src/runtime.js [hunk 2]
```diff
@@ -228,7 +214,54 @@ function adoptShared(global, callback) {
...
+      }
+    };
+
+    /**
+     * Registers an ampdoc service.
+     * @param {string} name
+     * @param {function(new:!Object, !./service/ampdoc-impl/AmpDoc)} opt_class
+     * @param {function(!./service/ampdoc-impl/AmpDoc):!Object} opt_factory
+     */
+    global.AMP.registerDocService = function(name, opt_class, opt_factory) {
```
Comment (cramforce) [inline src/runtime.js]: Not called yet?

File: src/service/modules-impl.js [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+  }
+
+  /**
+   * @param {string} name
+   * @return {!ModuleHolderDef}
+   * @private
+   */
+  getModuleHolder_(name) {
+    let holder = this.modules_[name];
+    if (!holder) {
```
Comment (cramforce) [inline src/service/modules-impl.js]: Does this structure leave this module?

File: src/runtime.js (hunk 1 above)
Comment (dvoytenko) [inline src/runtime.js]: So, it's backward compatible for extensions. Unfortunately, this is NOT backward compatible for runtime. E.g. if an older runtime loads a newer extension (runtime doesn't understand struct, but new extension is using it), we have an issue. Any advice? We could consider deploying just `fnOrStruct` support first and modify extension wrapper a week later?

File: src/service/modules-impl.js [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+      let resolve, reject;
+      const promise = new Promise((resolve_, reject_) => {
```
Comment (cramforce) [inline src/service/modules-impl.js]: Will every extension need such a promise?

File: src/runtime.js (hunk 2 above)
Comment (dvoytenko) [inline src/runtime.js]: Unfortunately, not yet ready to do services quite yet. They are still a bit of a web of dependencies. The only leaf extension I'm ready to migrate right now is amp access and, naturally, i'd like to do it a bit more carefully.

File: src/runtime.js (hunk 1 above)
Comment (cramforce) [inline src/runtime.js]: Yes, will need to go for that.

Comment (dvoytenko) [inline src/runtime.js]: Done

File: src/service/modules-impl.js (hunk 3 above)
Comment (dvoytenko) [inline src/service/modules-impl.js]: No. Never.

File: src/service/modules-impl.js (hunk 4 above)
Comment (dvoytenko) [inline src/service/modules-impl.js]: Redid to make this lazy.

File: src/runtime.js [hunk 5]
```diff
@@ -268,43 +281,171 @@ export function adopt(global) {
...
+  if (opt_css) {
+    addDocFactoryToExtension(extensions, ampdoc => {
+      installStylesForShadowRoot(ampdoc.getRootNode(), opt_css,
+          /* isRuntimeCss */ false, name);
+    });
+  }
+}
+
+
+/**
```
Comment (cramforce) [inline src/runtime.js]: Docs please

File: src/service/extensions-impl.js [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+
+  /**
+   * Installs all ampdoc factories previously registered with
+   * `addDocFactory_`.
+   * @param {!./ampdoc-impl.AmpDoc} ampdoc
+   * @param {!Array<string>} extensionIds
+   * @private
+   * @restricted
+   */
+  instrumentShadowDoc_(ampdoc, extensionIds) {
```
Comment (cramforce) [inline src/service/extensions-impl.js]: Could there be a better name? E.g. `installExtensionsInShadowDoc`?

File: src/service/extensions-impl.js [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+   * @private
+   */
+  getExtensionHolder_(extensionId) {
+    let holder = this.extensions_[extensionId];
+    if (!holder) {
+      const extension = {
+        elements: {},
+      };
+      holder = {
+        extension,
```
Comment (cramforce) [inline src/service/extensions-impl.js]: Also initialize undefined fields.

**Comment by cramforce** [comment]
Made a pass. Looks good.

File: src/runtime.js (hunk 5 above)
Comment (dvoytenko) [inline src/runtime.js]: Done

File: src/service/extensions-impl.js (hunk 6 above)
Comment (dvoytenko) [inline src/service/extensions-impl.js]: Done

File: src/service/extensions-impl.js (hunk 7 above)
Comment (dvoytenko) [inline src/service/extensions-impl.js]: Done

**Comment by dvoytenko** [comment]
@cramforce Tests have been completed. PTAL.

**Comment by cramforce** [comment]
LGTM

**Comment by cramforce** [comment]
LGTM

**Pull request merged**